controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -1377,166 +1389,314 @@ private TxnStatus handleDataNotFoundException(Throwable ex) {
+        Map<Integer, List<Long>> shards = sealedSegmentSizes.keySet().stream()
+                                                            .collect(Collectors.groupingBy(x -> StreamSegmentNameUtils.getSegmentNumber(x) / SHARD_SIZE));
+        return Futures.allOf(shards.entrySet().stream().map(x -> {
+            int shard = x.getKey();
+            List<Long> segments = x.getValue();
+
+            return createSealedSegmentSizeMapShardIfAbsent(shard).thenCompose(v -> getSealedSegmentSizesMapShardData(shard)
+                    .thenApply(y -> {
+                        SealedSegmentsMapShard mapShard = SealedSegmentsMapShard.fromBytes(y.getData());
+                        segments.forEach(z -> mapShard.addSealedSegmentSize(z, sealedSegmentSizes.get(z)));
+                        return updateSealedSegmentSizesMapShardData(shard, new Data(mapShard.toBytes(), y.getVersion()));
+                    }));
+        }).collect(Collectors.toList()));
     }
 
-    protected int getTransactionEpoch(UUID txId) {
-        // epoch == UUID.msb >> 32
-        return TableHelper.getTransactionEpoch(txId);
+    private Map<StreamSegmentRecord, Integer> convertToSpan(EpochRecord epochRecord) {
+        return epochRecord.getSegments().stream()
+                          .collect(Collectors.toMap(x -> x, x -> epochRecord.getEpoch()));
     }
 
-    abstract CompletableFuture<Void> deleteStream();
+    private Segment transform(StreamSegmentRecord segmentRecord) {
+        return new Segment(segmentRecord.segmentId(), segmentRecord.getCreationTime(),
+                segmentRecord.getKeyStart(), segmentRecord.getKeyEnd());
+    }
 
-    abstract CompletableFuture<CreateStreamResponse> checkStreamExists(final StreamConfiguration configuration, final long creationTime, final int startingSegmentNumber);
+    private List<Segment> transform(List<StreamSegmentRecord> segmentRecords) {
+        return segmentRecords.stream().map(this::transform).collect(Collectors.toList());
+    }
+
+    private Set<Segment> transform(Set<StreamSegmentRecord> segmentRecords) {
+        return segmentRecords.stream().map(this::transform).collect(Collectors.toSet());
+    }
+    
+    private CompletableFuture<List<EpochRecord>> fetchEpochs(int fromEpoch, int toEpoch) {
+        // fetch history time series chunk corresponding to from.
+        // read entries till either last entry or till to
+        // if to is not in this chunk fetch the next chunk and read till to
+        // keep doing this until all records till to have been read.
+        // keep computing history record from history time series by applying delta on previous.
+        return getActiveEpochRecord(false)
+                .thenApply(currentEpoch -> currentEpoch.getEpoch() / HISTORY_CHUNK_SIZE)
+                .thenCompose(latestChunkNumber -> Futures.allOfWithResults(
+                        IntStream.range(fromEpoch / HISTORY_CHUNK_SIZE, toEpoch / HISTORY_CHUNK_SIZE + 1)
```

Comment:
I'm not sure I understand how this range is being computed. Aren't `fromEpoch` and `toEpoch` likely to be small  numbers, which will give us a rangeof `(0, 1 )`, unless the epoch numbers are greater than `HISTORY_CHUNK_SIZE`? Maybe that's the intention, that for most streams, we will need only a single chunk...

Reply:
For a stream which scaled every second, we would have ~31M epochs in an year. 

These would be divided across 31k chunks. 

If someone wants to perform batch processing on such a stream and called getSegmentsBetween (streamCheckPointAtEndOfYear1, streamCheckPointAtEndOfYear2) we could have 31k chunks between fromEpoch and toEpoch.

Reply:
But yes, for most streams that do not scale that aggressively, the number of chunks over which we loop would be small.  

Aggressive scaling can typically happen if stream has very large number of segments and there is enough variance in the traffic across key ranges to frequently make segments hot or cold.